envoy/server.go

```diff
@@ -0,0 +1,437 @@
+package envoy
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"strconv"
+	"time"
+
+	"github.com/Nitro/sidecar/catalog"
+	"github.com/Nitro/sidecar/config"
+	"github.com/Nitro/sidecar/service"
+	api "github.com/envoyproxy/go-control-plane/envoy/api/v2"
+	core "github.com/envoyproxy/go-control-plane/envoy/api/v2/core"
+	endpoint "github.com/envoyproxy/go-control-plane/envoy/api/v2/endpoint"
+	listener "github.com/envoyproxy/go-control-plane/envoy/api/v2/listener"
+	route "github.com/envoyproxy/go-control-plane/envoy/api/v2/route"
+	hcm "github.com/envoyproxy/go-control-plane/envoy/config/filter/network/http_connection_manager/v2"
+	tcpp "github.com/envoyproxy/go-control-plane/envoy/config/filter/network/tcp_proxy/v2"
+	discovery "github.com/envoyproxy/go-control-plane/envoy/service/discovery/v2"
+	"github.com/envoyproxy/go-control-plane/pkg/cache"
+	xds "github.com/envoyproxy/go-control-plane/pkg/server"
+	"github.com/envoyproxy/go-control-plane/pkg/wellknown"
+	"github.com/gogo/protobuf/proto"
+	"github.com/golang/protobuf/ptypes"
+	"github.com/golang/protobuf/ptypes/duration"
+	"github.com/relistan/go-director"
+	log "github.com/sirupsen/logrus"
+	"google.golang.org/grpc"
+)
+
+const (
+	// !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+	// TODO: Move this code in a common module and de-duplicate in sidecarhttp/envoy_api.go
+	// !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+	// Used to join service name and port. Must not occur in service names
+	ServiceNameSeparator = ":"
+)
+
+type xdsCallbacks struct{}
+
+func (*xdsCallbacks) OnStreamOpen(context.Context, int64, string) error  { return nil }
+func (*xdsCallbacks) OnStreamClosed(int64)                               {}
+func (*xdsCallbacks) OnStreamRequest(int64, *api.DiscoveryRequest) error { return nil }
+func (*xdsCallbacks) OnStreamResponse(_ int64, req *api.DiscoveryRequest, _ *api.DiscoveryResponse) {
+	if req.GetErrorDetail().GetCode() != 0 {
+		log.Errorf("Received Envoy error code %d. Check Envoy logs for details.", req.GetErrorDetail().GetCode())
+		log.Debugf("Envoy error: %s", req.GetErrorDetail().GetMessage())
```

Comment:
It dawned on me that Sidecar wasn't printing anything in case Envoy rejected an update, so I added this code. Annoyingly, Envoy sends back only a string message (`req.GetErrorDetail().GetDetails()` is `nil`) with the entire chunk of invalid config as pretty-printed JSON which would make the log entry printed by `log.Errorf` multiline if I dump it there, hence the separate `log.Debugf`. Here's a sample:
```
ERRO[2020-01-04T23:39:51Z] Received Envoy error code 13. Check Envoy logs for details.
DEBU[2020-01-04T23:39:51Z] Envoy error: Error adding/updating cluster(s) whoami-http:10080: Field 'connect_timeout' is missing in: name: "whoami-http:10080"
type: STATIC
load_assignment {
  cluster_name: "whoami-http:10080"
  endpoints {
    lb_endpoints {
      endpoint {
        address {
          socket_address {
            address: "192.168.0.87"
            port_value: 32772
          }
        }
      }
    }
  }
}
```
I guess I could remove the newlines or I could extract only the first line of the message and add it to `log.Errorf`, if you think it's useful.

Also, Envoy doesn't have a backoff timer when getting invalid updates so it keeps trying over and over really fast, potentially flooding its logs as well as Sidecar's if such a situation occurs.

Reply:
I'd just remove all the carriage returns from this log line.